### PR TITLE
#4 send image/video

### PR DESCRIPTION
**Pull request by mukaiu**
Support Message Type of Image/Video for #4 

Add
wechaty.uploadMedia(stream: NodeJS.ReadableStream, filename: string, toUserName: string)
wechaty.sendMedia(message: Message)

### CODE REVIEW

**Comment by Samurais** [comment]
@zixia @lijiarui, @mukaiu has tested the functions. Please give some comments. Thanks.

**Comment by zixia** [comment]
@Samurais Great, I'll have a look into it.

@mukaiu Good job! 👍

**Comment by zixia** [comment]
Update: 

I realized that `contact.say('/tmp/file.jpg')` will not work in sudden, because that's a string...

Do you have any suggestion for using the file path name directly for this scanerio?

**Comment by lijiarui** [comment]
How about design a new type like 
```
export type Content = {
    Type?: 'file' | 'pic'
    Content: ''/tmp/file.jpg''
}

contact.say(para: Content | string)
```

**Comment by dcsan** [comment]
I tested out this PR using

```
git fetch origin pull/337/head
git checkout -b send-image-pr FETCH_HEAD
```


```
[bot test dc]<DC>:{TEXT}dc-alias:Ding
23:35:53 INFO Bot REPLY: img
TypeError: this.puppet.getBaseRequest is not a function
    at Wechaty.<anonymous> (/Users/dc/dev/xx/wechaty/src/wechaty.ts:458:41)
    at next (native)
    at fulfilled (/Users/dc/dev/xx/wechaty/dist/src/wechaty.js:4:58)
    at process._tickDomainCallback (internal/process/next_tick.js:129:7)
```

**Comment by mukaiu** [comment]
@dcsan 
This is running ding-img-bot.ts?

**Comment by dcsan** [comment]
@mukaiu yes

> ts-node example/ding-img-bot.ts

(run as an npm script)

**Comment by zixia** [comment]
@lijiarui Yes, a new type is needed with this scenario.

What I'm thinking about is to create a new Class for doing this instead of an Interface. I.e. `new File('/tmp/a.pdf')` because this class could set the MIME information of the file automatically instead of by hand, which will be more convenience.

@mukaiu @JasLin  I want to hear from you too, because this API will needed after this PR is merged. 

Thanks!

**Comment by mukaiu** [comment]
I think,
```
contact.say(MediaMessage) 
```

**Comment by zixia** [comment]
Yap, that's a better idea to solve this problem! :D
👍 @mukaiu

**Comment by mukaiu** [comment]
@zixia 
implement 
contact.say(new MediaMessage('/tmp/file.jpg'))
message.say(new MediaMessage('/tmp/haha.gif'))

but the logic of room.say is uncertain.

**Comment by mukaiu** [comment]
image
test error,help~~~

**Comment by dcsan** [comment]
Is there a docker image with this version of Wechaty? It would make testing much easier...

File: src/message-media.ts [hunk 1]
```diff
@@ -140,6 +147,9 @@ export class MediaMessage extends Message {
   }
 
   public filename(): string {
```
Comment (zixia) [inline src/message-media.ts]: According to <https://nodejs.org/api/path.html#path_path_parse_path>, I think we should rename `filename()` at here to `fileBase()`, return a full file base name(without path) like `music.mp3`

File: src/message-media.ts [hunk 2]
```diff
@@ -140,6 +147,9 @@ export class MediaMessage extends Message {
...
+    if (this.filepath)
```
Comment (zixia) [inline src/message-media.ts]: Change to:

```ts
if (this.fileName && this.fileExt) {
  return this.fileName + '.' + this.fileExt
}
```

File: src/message.ts [hunk 3]
```diff
@@ -492,31 +493,44 @@ export class Message implements Sayable {
     })
   }
 
-  public say(content: string, replyTo?: Contact|Contact[]): Promise<any> {
+  public say(content: string | MediaMessage, replyTo?: Contact|Contact[]): Promise<any> {
```
Comment (zixia) [inline src/message.ts]: Please use Override to declare `say()` at here.

```ts
say(textOrMedia: string | MediaMessage)
```

File: src/message.ts [hunk 4]
```diff
@@ -492,31 +493,44 @@ export class Message implements Sayable {
...
-      let mentionList
-      if (Array.isArray(replyTo)) {
-        m.to(replyTo[0])
-        mentionList = replyTo.map(c => '@' + c.name()).join(' ')
-      } else {
-        m.to(replyTo)
-        mentionList = '@' + replyTo.name()
       }
-      m.content(mentionList + ' ' + content)
+    } else {
```
Comment (zixia) [inline src/message.ts]: `} else if (textOrMedia instnaceof MediaMessage) {`

File: src/puppet-web/puppet-web.ts [hunk 5]
```diff
@@ -296,11 +303,114 @@ export class PuppetWeb extends Puppet {
...
+    let first = cookie.find(c => c.name === 'webwx_data_ticket')
+    let webwxDataTicket = first && first.value
+    let size = buffer.length
+
+    let uploadMediaRequest = {
+      BaseRequest: baseRequest,
+      FileMd5: md5,
+      FromUserName: this.self().id,
+      ToUserName: toUserName,
+      UploadType: 2,
```
Comment (zixia) [inline src/puppet-web/puppet-web.ts]: Could you change this magic number to a `enum` value?

File: src/puppet-web/puppet-web.ts [hunk 6]
```diff
@@ -296,11 +303,114 @@ export class PuppetWeb extends Puppet {
...
+      ClientMediaId: +new Date,
+      MediaType: 4,
```
Comment (zixia) [inline src/puppet-web/puppet-web.ts]: Could you change this magic number to a `enum` value?

File: src/puppet-web/puppet-web.ts [hunk 7]
```diff
@@ -296,11 +303,114 @@ export class PuppetWeb extends Puppet {
...
-    const content     = message.content()
+  private async uploadMedia( filepath: string, toUserName: string): Promise<string> {
+    if (!filepath)
+      throw new Error('no filepath')
+
+    let filename = path.basename(filepath)
+    let ext = path.extname(filepath)
+
+    let contentType = UtilLib.mime(ext)
+    let mediatype
```
Comment (zixia) [inline src/puppet-web/puppet-web.ts]: Could you define a new type for MediaType like:

```ts
type MediaTpye = 'pic' | 'video' | 'doc'
```

File: src/message-media.ts [hunk 8]
```diff
@@ -16,19 +16,26 @@ import {
...
   MsgType,
+  // MsgRawObj,
 }                     from './message'
 import { UtilLib }    from './util-lib'
 import { PuppetWeb }  from './puppet-web/puppet-web'
 import { Bridge }     from './puppet-web/bridge'
 
 export class MediaMessage extends Message {
   private bridge: Bridge
+  private filepath: string
```
Comment (mukaiu) [inline src/message-media.ts]: Replace the ReadStream with Buffer

**Comment by mukaiu** [comment]
@dcsan docker pull mukaiu/wechaty

File: src/puppet-web/puppet-web.ts (hunk 5 above)
Comment (mukaiu) [inline src/puppet-web/puppet-web.ts]: UploadType optional value 1 or 2, use unknown

**Comment by zixia** [comment]
Hi @mukaiu,

Thank you so much for this Pull Request!

Now I feel it's good enough for supporting attachment messages in Wechaty, and I believe it will help many developers who want to use this functionality!

I have no more reviews for your PR, and I'll approve the changes. 

However, We all know it's not perfect. I'm afraid there's some part that we missed to review. Could you talk with other _Reviewers_, and let some of them to _Approve Changes_? 

Let's do merge the PR after at least other two contributors approved it. Cheers!

**Comment by dcsan** [comment]
apologies for delay, I have been trying to check this out but haven't got a good docker usage development workflow, it takes too long to make a code change and test out, so I haven't been able to test it yet.

FWIW my flow is to build a container with mukai/wechaty fork as below
but then i have to rebuild the container, copy in my own code, run it... 
since its a new container each time I also have to scan a QR code each dev loop.
this takes a few mins each code change.

is there a faster dev workflow you use?

I guess I could just run off a local build of wechaty, but I had problems with googlechrome and other deps, so aiming to use docker.

perhaps I can just `exec bash` on the docker container and work directly within it, but not sure how that works yet with the `CMD [""]` and ENTRYPOINT here
https://hub.docker.com/r/zixia/wechaty/~/dockerfile/

any other suggestions for fast workflow appreciated.

```
## using images fork
FROM mukaiu/wechaty
EXPOSE 8080
RUN mkdir -p /bot
WORKDIR /bot
COPY . /bot

CMD ["index.js"]
```

**Comment by mukaiu** [comment]
@dcsan 
Try using Docker Volume to hang in the local directory to the docker bot directory.
```
docker run --name  bot -d -v LOCAL_DIR:/bot zixia/wechaty mybot.js
```
After modifying the code, restart container
```
docker restart bot
```

Use the following command to display the container log
```
docker logs -f bot
```

**Comment by dcsan** [comment]
@mukaiu thanks i didn't know about docker volumes. perhaps if i combine that with hot reloading it will be much less painful to develop this way.

So i went to try and test this. in a wechaty checkout

```
  git checkout master
  git pull
  git fetch origin pull/337/head\n
```

but now it seems the `ding-image` bot doesnt exist, and there's no demo code in the ding-dong or media-file bot. Is there a usage example anywhere?

**Comment by dcsan** [comment]
UPDATE

I tried just doing a clone of @mukaiu 's fork, and that seems to work very well in sending an image message. So at this point I'm not sure about this PR but the forked repo is functional.

https://github.com/mukaiu/wechaty/blob/master/example/ding-dong-bot.ts#L75

So, I guess my method for reviewing this PR isn't working.
Just FYI since I am not a maintainer of this repo, it's a bit more complicated to check out a fork.

this message signature extensions did not come down for me

```ts
  public async say(textOrMedia: string | MediaMessage): Promise<void> {
```

even though it shows up in the PR
https://github.com/Chatie/wechaty/pull/337/files#diff-91d335b29ba8ed8ebb1d29260610d794R584

```
  git checkout master
  git pull
  git fetch origin pull/337/head
```

update

  git pull origin pull/337/head

seems to do it >.<

**Comment by dcsan** [comment]
if you did make an "edge" branch it would make it a lot easier for me to make a fork based off of that and start modifying / sending PRs. Making a PR based on someone else's PR on a repo that you don't own is just too much complication.

**Comment by zixia** [comment]
Thanks @dcsan @lijiarui for approving this PR, it really helps.

@dcsan I like your idea of creating a _edge_ branch to automaticaly merge the Pull Requests(which passed CI) for testers.

Do you know any web service or tools to do this kind of job(merge pull requests to a branch) automaticaly? That will be very efficiency.

**Comment by dcsan** [comment]
I'm not sure, but I dont think we have *that* many PRs now?
i have seen some other github projects with very active bots, so I will try to take a note next time I see this. Rails and many other projects keep edge branch.

**Comment by dcsan** [comment]
https://github.com/bors-ng/bors-ng
https://github.com/calmh/mergebot

**Comment by jiang-wei** [comment]
@zixia any plan to release this merge to a new tag?
cannot wait to try this new feature.

**Pull request merged**